File: packages/deon-go/datasign.go
package deon

import (
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

// Typing a document against a declared contract (specification 14.1). The conservative typer of §14
// guesses from the value, so it must refuse whenever a guess could be wrong. A datasign contract is
// the other half: it supplies the intent the value cannot carry, and 007 becomes 7 exactly where a
// contract declared it a number, and nowhere else. This is an adapter to the datasign format, which
// belongs to its own project, and the rules below are that format's own.

const datasignSource = "<datasign>"

type datasignField struct {
	name     string
	declared string
	required bool
}

type signatures map[string][]datasignField

// sign applies the caller's contracts to an evaluated root (specification 14.1). Nothing happens
// without a datasign map: an empty map is the identity, and there is no reason to read a contract
// nobody is going to apply.
func sign(root Value, options *ParseOptions) Value {
	if len(options.DatasignMap) == 0 {
		return root
	}

	sigs := signatures{}
	for _, file := range options.DatasignFiles {
		source := readContract(file, options)
		for name, fields := range parseDatasign(source) {
			sigs[name] = fields
		}
	}

	return applyDatasign(root, sigs, options.DatasignMap)
}

// readContract reads a .datasign file. Reading one is filesystem access like any other and subject to
// §9: a raw string handed to Parse grants nothing, so a contract it names may not be read from a disk.
// A contract supplied in Resources needs no grant, because nothing is being reached.
func readContract(file string, options *ParseOptions) string {
	target := file
	if !filepath.IsAbs(file) && options.Filebase != "" {
		target = filepath.Join(options.Filebase, file)
	}

	if data, ok := options.Resources[target]; ok {
		return data
	}
	if data, ok := options.Resources[file]; ok {
		return data
	}

	if !options.AllowFilesystem {
		fail(CapabilityDenied,
			"Reading the datasign file '"+file+"' requires filesystem access.", headSpan(file))
	}
	bytes, err := os.ReadFile(target)
	if err != nil {
		fail(ResourceIO, "Unable to read the datasign file '"+file+"'.", headSpan(file))
	}
	return string(bytes)
}

// #region reading a contract
var datasignEntity = regexp.MustCompile(`^\s*data\s+(\w+)\s*\{`)

func parseDatasign(source string) signatures {
	sigs := signatures{}
	open := ""

	for _, line := range strings.Split(source, "\n") {
		trimmed := strings.TrimLeft(line, " \t")
		if strings.HasPrefix(trimmed, "//") || strings.HasPrefix(trimmed, "/*") ||
			strings.HasPrefix(trimmed, "*") || strings.HasPrefix(trimmed, "@") {
			continue
		}

		value := line
		if at := strings.Index(value, "//"); at >= 0 {
			value = value[:at]
		}
		if strings.TrimSpace(value) == "" {
			continue
		}

		if match := datasignEntity.FindStringSubmatch(value); match != nil {
			open = match[1]
			sigs[open] = []datasignField{}
			continue
		}
		if strings.HasPrefix(strings.TrimLeft(value, " \t"), "}") {
			open = ""
			continue
		}
		if open == "" {
			continue
		}

		colon := strings.Index(value, ":")
		if colon == -1 {
			continue
		}

		// A `?` anywhere on the line marks the field optional — datasign's own rule — and is removed
		// from both the name and the type, so `nickname?: string` and `nickname: string?` are one
		// declaration.
		optional := strings.Contains(value, "?")
		name := strings.ReplaceAll(strings.TrimSpace(value[:colon]), "?", "")
		declared := strings.ReplaceAll(strings.TrimSpace(strings.TrimRight(strings.TrimSpace(value[colon+1:]), ";")), "?", "")
		if name == "" || declared == "" {
			continue
		}

		sigs[open] = append(sigs[open], datasignField{name: name, declared: declared, required: !optional})
	}

	return sigs
}

// #endregion reading a contract

// #region numbers
var datasignDecimal = regexp.MustCompile(`^[+-]?([0-9]+\.?[0-9]*|\.[0-9]+)([eE][+-]?[0-9]+)?$`)

// datasignNumeric reads a string as a number the way ECMAScript's Number(string) does, which §14.1
// fixes as the grammar — wider than Go's own strconv, and different from it: Go parses "inf" and
// rejects "0x10", where this must reject the first and read the second as 16. So the grammar is
// written out rather than delegated to the host.
func datasignNumeric(text string) (float64, bool) {
	trimmed := strings.TrimSpace(text)
	if trimmed == "" {
		return 0, false
	}

	for _, radix := range []struct {
		prefix string
		base   int
	}{{"0x", 16}, {"0X", 16}, {"0o", 8}, {"0O", 8}, {"0b", 2}, {"0B", 2}} {
		if strings.HasPrefix(trimmed, radix.prefix) {
			n, err := strconv.ParseUint(trimmed[2:], radix.base, 64)
			if err != nil {
				return 0, false
			}
			return float64(n), true
		}
	}

	if !datasignDecimal.MatchString(trimmed) {
		return 0, false
	}
	n, err := strconv.ParseFloat(trimmed, 64)
	if err != nil || isInfOrNaN(n) {
		return 0, false
	}
	return n, true
}

// #endregion numbers

// #region applying a contract
func datasignDescribe(value Value) string {
	switch value.(type) {
	case string:
		return "a string"
	case []Value:
		return "a list"
	case *Map:
		return "a map"
	default:
		return "a value"
	}
}

// verbatim carries a value the contract said nothing about across unchanged. Emphatically not the
// typer, which guesses: a key the contract does not mention has not been declared to be anything.
func verbatim(value Value) any {
	switch v := value.(type) {
	case []Value:
		out := make([]any, len(v))
		for i, item := range v {
			out[i] = verbatim(item)
		}
		return out
	case *Map:
		out := NewMap()
		for _, key := range v.Keys() {
			member, _ := v.Get(key)
			out.Set(key, verbatim(member))
		}
		return out
	default:
		return value
	}
}

func typeDatasign(value Value, declared string, sigs signatures, path string) any {
	declared = strings.TrimSpace(declared)

	if strings.HasSuffix(declared, "[]") {
		list, ok := value.([]Value)
		if !ok {
			fail(TypeMismatch, "Expected '"+path+"' to be a list for '"+declared+"', found "+datasignDescribe(value)+".", headSpan(datasignSource))
		}
		item := strings.TrimSpace(declared[:len(declared)-2])
		out := make([]any, len(list))
		for i, entry := range list {
			out[i] = typeDatasign(entry, item, sigs, path+"["+strconv.Itoa(i)+"]")
		}
		return out
	}

	if declared == "string" || declared == "number" || declared == "boolean" {
		text, ok := value.(string)
		if !ok {
			fail(TypeMismatch, "Expected '"+path+"' to be a string for '"+declared+"', found "+datasignDescribe(value)+".", headSpan(datasignSource))
		}
		switch declared {
		case "string":
			return text
		case "boolean":
			if text == "true" {
				return true
			}
			if text == "false" {
				return false
			}
			fail(TypeMismatch, "Expected '"+path+"' to be 'true' or 'false' for 'boolean', found '"+text+"'.", headSpan(datasignSource))
		default:
			n, ok := datasignNumeric(text)
			if !ok {
				fail(TypeMismatch, "Expected '"+path+"' to be a number, found '"+text+"'.", headSpan(datasignSource))
			}
			return n
		}
	}

	entity, ok := sigs[declared]
	if !ok {
		// A type defined somewhere else; a value is not guessed at merely because its type was not
		// found.
		return verbatim(value)
	}

	container, ok := value.(*Map)
	if !ok {
		fail(TypeMismatch, "Expected '"+path+"' to be a map for '"+declared+"', found "+datasignDescribe(value)+".", headSpan(datasignSource))
	}

	fields := map[string]datasignField{}
	for _, field := range entity {
		fields[field.name] = field
	}

	out := NewMap()
	for _, key := range container.Keys() {
		member, _ := container.Get(key)
		if field, declared := fields[key]; declared {
			out.Set(key, typeDatasign(member, field.declared, sigs, path+"."+key))
		} else {
			out.Set(key, verbatim(member))
		}
	}
	for _, field := range entity {
		if field.required && !container.Has(field.name) {
			fail(TypeMismatch, "Required field '"+path+"."+field.name+"' of '"+declared+"' is missing.", headSpan(datasignSource))
		}
	}
	return out
}

func applyDatasign(root Value, sigs signatures, mapping map[string]string) any {
	if len(mapping) == 0 {
		return verbatim(root)
	}

	container, ok := root.(*Map)
	if !ok {
		fail(TypeMismatch, "A datasign map requires a root map, found "+datasignDescribe(root)+".", headSpan(datasignSource))
	}

	out := NewMap()
	for _, key := range container.Keys() {
		member, _ := container.Get(key)
		if declared, named := mapping[key]; named {
			out.Set(key, typeDatasign(member, declared, sigs, key))
		} else {
			out.Set(key, verbatim(member))
		}
	}
	return out
}

// #endregion applying a contract

File: packages/deon-java/source/deon/Datasign.java
package deon;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.OptionalDouble;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Typing a document against a declared contract (specification 14.1). The conservative typer guesses
 * from the value and so refuses whenever a guess could be wrong; a datasign contract is the other half —
 * it supplies the intent the value cannot carry, so {@code 007} becomes 7 exactly where a contract
 * declared it a number, and nowhere else. This is an adapter to the datasign format, whose rules are its
 * own.
 */
final class Datasign {
    private static final String SOURCE = "<datasign>";

    private record Field(String name, String declared, boolean required) {
    }

    private final ParseOptions options;
    private final Map<String, List<Field>> signatures = new LinkedHashMap<>();

    private Datasign(ParseOptions options) {
        this.options = options;
    }

    static Object sign(Object root, ParseOptions options) {
        if (options.datasignMap.isEmpty()) {
            return root;
        }
        Datasign datasign = new Datasign(options);
        for (String file : options.datasignFiles) {
            datasign.parseContract(datasign.readContract(file));
        }
        return datasign.apply(root);
    }

    // #region reading a contract
    private String readContract(String file) {
        String target = file;
        if (!file.startsWith("/") && options.filebase != null && !options.filebase.isEmpty()) {
            target = options.filebase + "/" + file;
        }
        if (options.resources.containsKey(target)) {
            return options.resources.get(target);
        }
        if (options.resources.containsKey(file)) {
            return options.resources.get(file);
        }
        if (!options.allowFilesystem) {
            throw new DeonException(Code.CAPABILITY_DENIED,
                    "Reading the datasign file '" + file + "' requires filesystem access.", Span.head(SOURCE));
        }
        byte[] bytes;
        try {
            bytes = Files.readAllBytes(Path.of(target));
        } catch (IOException e) {
            throw new DeonException(Code.RESOURCE_IO, "Unable to read the datasign file '" + file + "'.", Span.head(SOURCE));
        }
        if (!Interpreter.isValidUtf8(bytes)) {
            throw new DeonException(Code.RESOURCE_FORMAT, "The datasign file '" + file + "' is not valid UTF-8.", Span.head(SOURCE));
        }
        return new String(bytes, StandardCharsets.UTF_8);
    }

    private static final Pattern ENTITY = Pattern.compile("^\\s*data\\s+(\\w+)\\s*\\{");

    private void parseContract(String source) {
        List<Field> open = null;
        for (String line : source.split("\n", -1)) {
            String trimmed = line.stripLeading();
            if (trimmed.startsWith("//") || trimmed.startsWith("/*") || trimmed.startsWith("*") || trimmed.startsWith("@")) {
                continue;
            }
            String value = line;
            int comment = value.indexOf("//");
            if (comment >= 0) {
                value = value.substring(0, comment);
            }
            if (value.strip().isEmpty()) {
                continue;
            }
            Matcher match = ENTITY.matcher(value);
            if (match.find()) {
                open = new ArrayList<>();
                signatures.put(match.group(1), open);
                continue;
            }
            if (value.stripLeading().startsWith("}")) {
                open = null;
                continue;
            }
            if (open == null) {
                continue;
            }
            int colon = value.indexOf(':');
            if (colon < 0) {
                continue;
            }
            // A `?` anywhere on the line marks the field optional and is removed from both the name and
            // the type, so `nickname?: string` and `nickname: string?` are one declaration.
            boolean optional = value.contains("?");
            String name = value.substring(0, colon).strip().replace("?", "");
            String declared = stripTrailingSemicolon(value.substring(colon + 1).strip()).replace("?", "");
            if (name.isEmpty() || declared.isEmpty()) {
                continue;
            }
            open.add(new Field(name, declared, !optional));
        }
    }

    private static String stripTrailingSemicolon(String s) {
        String t = s.strip();
        return t.endsWith(";") ? t.substring(0, t.length() - 1).strip() : t;
    }
    // #endregion

    // #region numbers — ECMAScript Number(string), which section 14.1 fixes as the grammar
    private static final Pattern DECIMAL = Pattern.compile("^[+-]?([0-9]+\\.?[0-9]*|\\.[0-9]+)([eE][+-]?[0-9]+)?$");

    private static OptionalDouble numeric(String text) {
        String trimmed = text.strip();
        if (trimmed.isEmpty()) {
            return OptionalDouble.empty();
        }
        String[][] radixes = {{"0x", "16"}, {"0X", "16"}, {"0o", "8"}, {"0O", "8"}, {"0b", "2"}, {"0B", "2"}};
        for (String[] radix : radixes) {
            if (trimmed.startsWith(radix[0])) {
                try {
                    long n = Long.parseUnsignedLong(trimmed.substring(2), Integer.parseInt(radix[1]));
                    return OptionalDouble.of((double) n);
                } catch (NumberFormatException e) {
                    return OptionalDouble.empty();
                }
            }
        }
        if (!DECIMAL.matcher(trimmed).matches()) {
            return OptionalDouble.empty();
        }
        try {
            double n = Double.parseDouble(trimmed);
            return Double.isFinite(n) ? OptionalDouble.of(n) : OptionalDouble.empty();
        } catch (NumberFormatException e) {
            return OptionalDouble.empty();
        }
    }
    // #endregion

    // #region applying a contract
    private static String describe(Object value) {
        if (value instanceof String) {
            return "a string";
        }
        if (value instanceof List<?>) {
            return "a list";
        }
        if (value instanceof DeonMap) {
            return "a map";
        }
        return "a value";
    }

    private static DeonException mismatch(String message) {
        return new DeonException(Code.TYPE_MISMATCH, message, Span.head(SOURCE));
    }

    private Object typeDatasign(Object value, String declared, String path) {
        declared = declared.strip();

        if (declared.endsWith("[]")) {
            if (!(value instanceof List<?> list)) {
                throw mismatch("Expected '" + path + "' to be a list for '" + declared + "', found " + describe(value) + ".");
            }
            String item = declared.substring(0, declared.length() - 2).strip();
            List<Object> out = new ArrayList<>();
            for (int i = 0; i < list.size(); i++) {
                out.add(typeDatasign(list.get(i), item, path + "[" + i + "]"));
            }
            return out;
        }

        if (declared.equals("string") || declared.equals("number") || declared.equals("boolean")) {
            if (!(value instanceof String text)) {
                throw mismatch("Expected '" + path + "' to be a string for '" + declared + "', found " + describe(value) + ".");
            }
            return switch (declared) {
                case "string" -> text;
                case "boolean" -> {
                    if (text.equals("true")) {
                        yield Boolean.TRUE;
                    }
                    if (text.equals("false")) {
                        yield Boolean.FALSE;
                    }
                    throw mismatch("Expected '" + path + "' to be 'true' or 'false' for 'boolean', found '" + text + "'.");
                }
                default -> {
                    OptionalDouble n = numeric(text);
                    if (n.isEmpty()) {
                        throw mismatch("Expected '" + path + "' to be a number, found '" + text + "'.");
                    }
                    yield n.getAsDouble();
                }
            };
        }

        List<Field> entity = signatures.get(declared);
        if (entity == null) {
            return value; // a type defined elsewhere; a value is not guessed at
        }

        if (!(value instanceof DeonMap container)) {
            throw mismatch("Expected '" + path + "' to be a map for '" + declared + "', found " + describe(value) + ".");
        }

        Map<String, Field> fields = new LinkedHashMap<>();
        for (Field field : entity) {
            fields.put(field.name(), field);
        }
        DeonMap out = new DeonMap();
        for (String key : container.keys()) {
            Field field = fields.get(key);
            if (field != null) {
                out.set(key, typeDatasign(container.get(key), field.declared(), path + "." + key));
            } else {
                out.set(key, container.get(key)); // verbatim
            }
        }
        for (Field field : entity) {
            if (field.required() && !container.has(field.name())) {
                throw mismatch("Required field '" + path + "." + field.name() + "' of '" + declared + "' is missing.");
            }
        }
        return out;
    }

    private Object apply(Object root) {
        if (!(root instanceof DeonMap container)) {
            throw mismatch("A datasign map requires a root map, found " + describe(root) + ".");
        }
        DeonMap out = new DeonMap();
        for (String key : container.keys()) {
            String declared = options.datasignMap.get(key);
            if (declared != null) {
                out.set(key, typeDatasign(container.get(key), declared, key));
            } else {
                out.set(key, container.get(key)); // verbatim
            }
        }
        return out;
    }
    // #endregion
}
